Custom mapping

jrodriguez-rc opened the pull request:
Continue #41

Interface zif_ajson_custom_mapping for custom mapping

jrodriguez-rc commented:
I created the interface `zif_ajson_custom_mapping` with two methods: `to_abap` and `to_json`

@sbcgua What do you think about it?

sbcgua commented:
Thanks for the code. But many comments here. might even need a second review too. 
And I also merged the "keep_item_order" feature, needed for abapGit data serialization. So now there are some conflicts, sorry for this. But next merge is yours :)

sbcgua commented on src/mapping/package.devc.xml:
```diff
@@ -0,0 +1,10 @@
+ <asx:abap xmlns:asx="http://www.sap.com/abapxml" version="1.0">
+  <asx:values>
+   <DEVC>
+    <CTEXT>AJSON Mapping</CTEXT>
```
I'd prefer if there are no subpackages. Would like to keep it lean. And also may raise some issues in auto import procedure for abapgit. 
@larshp what do you think ? btw, maybe it makes sense to exclude some classes (or include only specific ones) in the auto transition process ?

jrodriguez-rc replied:
Ok, I'll leave only the main package

sbcgua commented on src/mapping/zcl_ajson_mapping_base.clas.abap:
```diff
@@ -0,0 +1,81 @@
+class zcl_ajson_mapping_base definition
```
OK, inheritance is something I'd avoid. It would make very difficult to change this class in future, if we want to.  E.g. the `cl_abap_unit_assert` vs `cl_aunit_assert` and stories like that.

I'd prefer pure composition. 

Also, to discuss it here, what if we combine all "standard" mappers in one class ? I mean they can be implemented as local classes to reduce number of global classes needed ? Instances can be created by public class methods. Not sure, just a suggestion. It is the same way I followed with main ajson class. Because all cases here are kind of "typical".

jrodriguez-rc replied:
The field mapping will apply to all the other types, I can call to the field mapping class instead of using as superclass

I created a class by mapping type because can have specific parameters, like camel case to set the first character as upper or lower case

sbcgua commented on src/mapping/zcl_ajson_mapping_base.clas.abap:
```diff
@@ -0,0 +1,81 @@
+      end of ty_mapping_field_ts,
+      ty_mapping_fields_tt type sorted table of ty_mapping_field_ts
+        with unique key sap
+        with unique sorted key json components json.
```
Pls move types to interface or make private. Also the naming is a bit strange, `ty_mapping_field_ts` - why "ts" suffix ? I'd recognize it as a sorted table. I'd suggest `ty_mapping_field` and then `ty_mapping_field_ts` for table type (instead of `tt`).
By the way maybe both `tt` and `ts` type needed - normal table for params, sorted for class attr.

sbcgua commented on src/mapping/zcl_ajson_mapping_base.clas.abap:
```diff
@@ -0,0 +1,81 @@
+    if sy-subrc = 0.
+      rv_result = ls_mapping_field-sap.
+    else.
+      rv_result = iv_segment.
```
How should be missing mapping handled ? What is the "contract" ? Should it return the passed node name ? Probably yes, but ! We cannot control it in really custom mappers. So the calling code must be aware of potential inaccuracy of a mapper. So maybe, i would return empty string if no mapping found. For discussion

jrodriguez-rc replied:
The idea is that the mapper returns the field name directly. If you prefer I can change it to return an empty value and check after calling this method

sbcgua commented on src/mapping/zcl_ajson_mapping_base.clas.abap:
```diff
@@ -0,0 +1,81 @@
+    if sy-subrc = 0.
+      rv_result = ls_mapping_field-json.
+    else.
+      rv_result = is_prefix-name.
```
same here

sbcgua commented on src/mapping/zcl_ajson_mapping_camel.clas.abap:
```diff
@@ -0,0 +1,82 @@
+
+  method zif_ajson_custom_mapping~to_json.
+
+    data lt_tokens type standard table of char256.
```
standard types pls. `c length 256` Here and in other places

sbcgua commented on src/mapping/zcl_ajson_mapping_to_lower.clas.abap:
```diff
@@ -0,0 +1,36 @@
+    data lt_tokens type standard table of char256.
+    field-symbols <token> like line of lt_tokens.
+
+    rv_result = super->to_json( is_prefix ).
```
Hmmmm ... custom mapping + upper/lower case at the same time ? what is the use case ?

I'd suggest to keep things separate. 

OR, if there is a real usecase (if!). Then I'd suggest implementing chains. A possibility to add multiple mappers one after another. But this may add some other challenges so maybe keep it for further improvement.

jrodriguez-rc replied:
I found some cases where the json field is longer than 30 characters, that's why I have the field mapping in all the mappings
![image](https://user-images.githubusercontent.com/56556686/104163399-a9c12600-53f6-11eb-8e60-85392b1b4c11.png)

sbcgua commented on src/mapping/zcl_ajson_mapping_to_upper.clas.abap:
```diff
@@ -0,0 +1,36 @@
+    data lt_tokens type standard table of char256.
+    field-symbols <token> like line of lt_tokens.
+
+    rv_result = super->to_json( is_prefix ).
```
same here

sbcgua commented on src/mapping/zif_ajson_custom_mapping.intf.abap:
```diff
@@ -0,0 +1,18 @@
+interface zif_ajson_custom_mapping
```
Suggest renaming to `zif_ajson_field_mapping` (or mapper). Because at this level of abstraction (interface) it is not yet custom, it is just a field mapping. While "custom" confuses what is it about. 
While param names in `parse` and `create_empty` are ok - in that context it is already "custom" - custom instance of `field_mapping`.
Sorry for some nerdiness ... :)

jrodriguez-rc replied:
No problem, you're right here, the name of the interface is not according to the purpose of it

sbcgua commented on src/mapping/zif_ajson_custom_mapping.intf.abap:
```diff
@@ -0,0 +1,18 @@
+    importing
+      !is_prefix       type zif_ajson=>ty_path_name
+    returning
+      value(rv_result) type string.
```
I'd suggest keeping similar interface for both methods, namely `iv_path` and `iv_name`. 

`to_abap` does not require segment, I mean `iv_name` is segment
`to_json` - also destructure into path and name. it is cleaner in the calling code, but in interface implementation if would be more readable to have separate params imho

sbcgua commented on src/zcl_ajson.clas.locals_imp.abap:
```diff
@@ -585,6 +590,7 @@ class lcl_json_to_abap implementation.
   method find_loc.
 
     data lt_path type string_table.
+    data lv_seg like line of lt_path.
```
Is lv_seg needed? Maybe it is ok to keep just `<seg>` ? (not sure, just a suggestion)

jrodriguez-rc replied:
I created it to have a trace, and leave the original values in `lt_path` table. I will remove `lv_seg` and use `<seg>`

sbcgua commented on src/zcl_ajson.clas.locals_imp.abap:
```diff
@@ -604,6 +610,15 @@ class lcl_json_to_abap implementation.
+            mi_custom_mapping->to_abap(
+                iv_path    = iv_path
+                iv_name    = iv_name
+                iv_segment = <seg> ).
```
seems like `iv_name` not needed in this method, see interface comment above.
`iv_path = lv_trace, iv_name  = <seg>`

sbcgua replied:
Also see the comment about returned param (potentially empty)

sbcgua commented on src/zcl_ajson.clas.locals_imp.abap:
```diff
@@ -604,6 +610,15 @@ class lcl_json_to_abap implementation.
+                iv_segment = <seg> ).
+      endif.
+
       <seg> = to_upper( <seg> ).
```
lv_seg is assigned above, but `<seg>` is upper cased. A bug ? Also pls check if it is possible to get rid of lv_seg at all (maybe)

jrodriguez-rc replied:
Yes, a bug, it is fixed, sorry for that

sbcgua commented on src/zcl_ajson.clas.locals_imp.abap:
```diff
@@ -890,6 +910,10 @@ class lcl_abap_to_json implementation.
     <n>-index = iv_index.
 
+    if mi_custom_mapping is bound.
+      <n>-name = mi_custom_mapping->to_json( is_prefix ).
```
what if empty value is returned ?

jrodriguez-rc replied:
I will change the behavior of the mappers and this, to leave the original name if the mapper returns an empty value

jrodriguez-rc replied:
hmmm, I need to use the `lv_seg` to get the value of `<seg>` if `to_abap` returns an empty value

sbcgua reviewed (changes requested):
Some comments above. Good code in general, but needs some reshaping. Sorry for some nerdiness :)) The feature is cool. Potentially might need a second review pass.

jrodriguez-rc commented:
@sbcgua the only comment pending is the use of field mapping in the other mappers

sbcgua commented:
OK, I think it is almost there :) 3 points:

1) "field mapping in the other mappers" - I think current approach is OK. mapper chains ... could be nice but probably overcomplicated.
2) I have a bad gut feeling about current `ty_mapping_fields`. I think it should have a standard table variant. And params should use it, instead sorted table. Because it limits the ways how you create this param (insert only). So imho `ty_mapping_fields` is for class attr only.
3) And then let's discuss about global/local classes :) I think the local classes are really underestimated. You add 4 more (small) global classes. While there are just 2 of them now (well 3 with the exception). This clutters the package a bit, at least for readability. What I'd suggest is adding one class, let's say `zcl_ajson_field_mapping` or `zcl_ajson_mapping_lib` or `zcl_ajson_std_mappings` and define several constructor methods:

```abap

class-methods camel_case " or create_camel_case_mapper ?
    importing
      it_mapping_fields   type zif_ajson_field_mapping~ty_mapping_fields optional
      iv_first_json_upper type abap_bool default abap_true
  returning
    value(ri_mapping) type ref to zif_ajson_field_mapping.
class-methods upper_case
    importing
      it_mapping_fields   type zif_ajson_field_mapping~ty_mapping_fields optional
  returning
    value(ri_mapping) type ref to zif_ajson_field_mapping.
class-methods lower_case
    importing
      it_mapping_fields   type zif_ajson_field_mapping~ty_mapping_fields optional
  returning
    value(ri_mapping) type ref to zif_ajson_field_mapping.
class-methods free_mapping " or custom_mapping
    importing
      it_mapping_fields   type zif_ajson_field_mapping~ty_mapping_fields
  returning
    value(ri_mapping) type ref to zif_ajson_field_mapping.

```

and then move all the current classes to locals without change (maybe just remove aliases - they will no be needed). And they all should be final.
What do you think ?

sbcgua commented:
btw, could I ask you to add some docs and examples after the code is finalized? (maybe a separate PR would do)

sbcgua commented:
closes #11

jrodriguez-rc commented:
I moved the classes to local classes of main `zcl_ajson_mapping`

About the type `ty_mapping_fields`, I changed it to a standard table, but with sorted keys. This allows to change data but won't allow having duplicated keys.

For the current abaplint errors, @larshp could be an issue?

I'm having `Statement does not exist in ABAPv702(or a parser error), "types"abaplint(parser_error)` at this sentence:

![image](https://user-images.githubusercontent.com/56556686/104297737-7f896a00-54c3-11eb-80ac-5c9c9f0f437a.png)
```
  types:
    begin of ty_mapping_field,
      abap type string,
      json type string,
    end of ty_mapping_field,
    ty_mapping_fields type standard table of ty_mapping_field with empty key
    with unique sorted key abap components abap
    with unique sorted key json components json.
```

jrodriguez-rc commented:
@larshp I found the problem. The check is ok, not the message, the error was from `with empty key instead of `types`.

sbcgua reviewed (changes requested):
From the code perspective look good now :) But pls check UTs, they are failing. Not sure why, but they work in master

![image](https://user-images.githubusercontent.com/15635498/104342960-0f4f0880-5504-11eb-832b-7571f3e521e3.png)

jrodriguez-rc commented:
There was an error converting `to_abap` with the `lv_seg`. Tests are now ok.
![image](https://user-images.githubusercontent.com/56556686/104428901-4ca7aa80-5585-11eb-8003-459b62352482.png)

sbcgua commented:
Yeah, that's why I like field symbols :)

sbcgua commented:
So, I guess this is good now. And it's a cool feature really. Thanks a lot for contrib ! :+1: :+1:
Merging.

P.S. If you have some spare time, could I ask you to add some docs for it in a separate PR. And maybe a test for lowercaseStartingCamelCase (seems that it's missing)

sbcgua commented:
@larshp is this feature needed in abapGit by the way ? I wonder if it makes sense to filter out classes within mirroring. Or maybe not.

larshp commented:
hmm, will probably keep it as is, for transparency

jrodriguez-rc commented:
Ready the test for camel case with first lower.

I will prepare some documentation and test cases, probably this weekend.

sbcgua commented:
:+1:
